libcodechecker/libhandlers/store.py

```diff
@@ -215,71 +206,53 @@ def collect_file_hashes_from_plist(plist_file):
-            else:
-                _, _, files = next(os.walk(input_path), ([], [], []))
+        if os.path.isfile(input_path):
+            files = [input_path]
+        else:
+            _, _, files = next(os.walk(input_path), ([], [], []))
 
+        with zipfile.ZipFile(zip_file, 'a') as zipf:
             for f in files:
                 plist_file = os.path.join(input_path, f)
                 if f.endswith(".plist"):
                     if collect_file_hashes_from_plist(plist_file):
                         LOG.debug(
                             "Copying file '{0}' to ZIP assembly dir..."
                             .format(plist_file))
-                        shutil.copy(plist_file, report_dir)
+                        zipf.write(plist_file, os.path.join('reports', f))
                     else:
                         LOG.warning("Skipping '{0}' because it contains "
                                     "a missing source file."
                                     .format(plist_file))
 
-        necessary_hashes = client.getMissingContentHashes(hash_to_file.keys())
+    necessary_hashes = client.getMissingContentHashes(hash_to_file.keys())
+    with zipfile.ZipFile(zip_file, 'a') as zipf:
         for f, h in file_to_hash.items():
             if h in necessary_hashes:
-                target_file = os.path.join(source_dir, f.lstrip('/'))
-
-                try:
-                    os.makedirs(os.path.dirname(target_file))
-                except os.error:
-                    # Directory already exists.
-                    pass
-
                 LOG.debug("File contents for '{0}' needed by the server"
                           .format(f))
-                shutil.copyfile(f, target_file)
-
-        hashes_fname = "content_hashes.json"
-        hashes_fpath = os.path.join(temp_dir, hashes_fname)
-        with open(hashes_fpath, "w") as f:
-            json.dump(file_to_hash, f)
-
-        with zipfile.ZipFile(zip_file, 'w') as zipf:
-            for root, _, files in os.walk(temp_dir):
-                for sf in files:
-                    f = os.path.join(root, sf)
-                    LOG.debug("[ZIP] Writing file '{0}' into mass store ZIP."
-                              .format(f))
-                    zipf.write(f, f.replace(temp_dir, ''))
-
-        # Compressing .zip file
-        with open(zip_file, 'rb') as source:
-            compressed = zlib.compress(source.read(),
-                                       zlib.Z_BEST_COMPRESSION)
-
-        with open(zip_file, 'wb') as target:
-            target.write(compressed)
-
-        LOG.debug("[ZIP] Mass store zip written at '{0}'".format(zip_file))
-    finally:
-        shutil.rmtree(temp_dir)
+
+                zipf.write(f, os.path.join('root', f.lstrip('/')))
+
+    with zipfile.ZipFile(zip_file, 'a') as zipf:
+        zipf.writestr('content_hashes.json', json.dumps(file_to_hash))
+
+    # Compressing .zip file
+    with open(zip_file, 'rb') as source:
```

Comment:
There is no way to write out compressed files in the first place?

Reply:
Yes, there is, but as far as I know if there is no way to set the compression rate.

Reply:
I think you can [set the compression](https://docs.python.org/2/library/zipfile.html#zipfile-objects).
` class zipfile.ZipFile(file[, mode[, compression[, allowZip64]]])`

Reply:
Here you can state whether a compression should happen, but the compression rate can't be set. And the default is not the best compression.